google-cloud-firestore/src/main/java/com/google/cloud/firestore/Query.java

```diff
@@ -1757,6 +1762,103 @@ public ApiFuture<QuerySnapshot> get() {
     return get(null);
   }
 
+  ApiFuture<QueryProfileInfo<QuerySnapshot>> getQueryProfileInfo(QueryMode queryMode) {
+    final SettableApiFuture<QueryProfileInfo<QuerySnapshot>> result = SettableApiFuture.create();
+
+    RunQueryRequest.Builder request = RunQueryRequest.newBuilder();
+    request
+        .setStructuredQuery(buildQuery())
+        .setParent(options.getParentPath().toString())
+        .setMode(queryMode);
+
+    final List<QueryDocumentSnapshot> documentSnapshots = new ArrayList<>();
+
+    ResponseObserver<RunQueryResponse> observer =
+        new ResponseObserver<RunQueryResponse>() {
+          Timestamp readTime;
+          Struct planStruct = Struct.getDefaultInstance();
+          Struct statsStruct = Struct.getDefaultInstance();
+
+          @Override
+          public void onStart(StreamController streamController) {}
+
+          @Override
+          public void onResponse(RunQueryResponse response) {
+            if (response.hasDocument()) {
+              Document document = response.getDocument();
+              QueryDocumentSnapshot documentSnapshot =
+                  QueryDocumentSnapshot.fromDocument(
+                      rpcContext, Timestamp.fromProto(response.getReadTime()), document);
+              documentSnapshots.add(documentSnapshot);
+            }
+
+            if (readTime == null) {
+              readTime = Timestamp.fromProto(response.getReadTime());
+            }
+
+            if (response.hasStats()) {
+              if (response.getStats().hasQueryPlan()) {
+                planStruct = response.getStats().getQueryPlan().getPlanInfo();
+              }
+              if (response.getStats().hasQueryStats()) {
+                statsStruct = response.getStats().getQueryStats();
+              }
+            }
+          }
+
+          @Override
+          public void onError(Throwable throwable) {
+            // When regular queries fail, the SDK uses the last received document as cursor and
+            // tries to
+            // resume querying after that cursor if the error is considered retryable.
+            // We should not do this when profiling a query.
+            result.setException(throwable);
+          }
+
+          @Override
+          public void onComplete() {
+            result.set(
+                new QueryProfileInfo<>(
+                    UserDataConverter.decodeStruct(planStruct),
+                    UserDataConverter.decodeStruct(statsStruct),
+                    QuerySnapshot.withDocuments(Query.this, readTime, documentSnapshots)));
+          }
+        };
+
+    rpcContext.streamRequest(request.build(), observer, rpcContext.getClient().runQueryCallable());
+
+    return result;
+  }
+
+  /**
+   * Performs the planning stage of this query, without actually executing the query. Returns an
+   * ApiFuture that will be resolved with the result of the query planning information.
```

Comment:
Is this an acceptable simplification? If so, can we use this phrasing elsewhere?

```suggestion
   * ApiFuture that will be resolved with query plan.
```